### PR TITLE
fix(tile): support in-place T.tile.silu/sigmoid (dst==src) on both backends

### LINKED ISSUE #1658: [Bug] T.tile.silu in-place (dst==src) silently produces wrong results on both ascendc and pto
## Bug Description

`T.tile.silu` 原地运算（dst == src 为同一 buffer）时静默产生错误结果，两后端均如此，无任何报错。已在 `request/说明和审查/框架.md` 第 12 条记录，根因已定位到 CANN 实现。

## Environment

**Hardware:**
- NPU Device: Ascend910B3（A2，dav-2201）

**Software:**
- OS: Linux
- CANN version: 8.5.2
- torch-npu version: 2.7.1
- Python version: 3.11
- TileLang-Ascend version: 0.1.4（refine-tile-relu-silu-sigmoid 分支）
- Installation method: build from source

## Code to Reproduce

```python
# Minimal reproducible code
import torch
import tilelang
import tilelang.language as T

PASS_CONFIGS = {
    tilelang.PassConfigKey.TL_ASCEND_AUTO_CV_COMBINE: True,
    tilelang.PassConfigKey.TL_ASCEND_AUTO_CV_SYNC: True,
    tilelang.PassConfigKey.TL_ASCEND_AUTO_SYNC: True,
    tilelang.PassConfigKey.TL_ASCEND_MEMORY_PLANNING: True,
}

@T.prim_func
def silu_inplace(
    A: T.Tensor((64, 64), "float16"),
    B: T.Tensor((64, 64), "float16"),
):
    with T.Kernel(1, is_npu=True) as (cid, vid):
        a_ub = T.alloc_ub((64, 64), "float16")
        T.copy(A, a_ub)
        T.tile.silu(a_ub, a_ub)  # in-place: dst == src
        T.copy(a_ub, B)

if __name__ == "__main__":
    a = torch.randn(64, 64, dtype=torch.float16, device="npu")
    for target in ["ascendc", "pto"]:
        compiled = tilelang.compile(silu_inplace, out_idx=[-1], pass_configs=PASS_CONFIGS, target=target)
        torch.npu.synchronize()
        b = compiled(a)
        torch.npu.synchronize()
        diff = (b - torch.nn.functional.silu(a)).abs().max().item()
        print(f"[silu-inplace] {target} max_diff={diff:.6f}")
```

## Error Message

无显式报错（静默错误）。实测数值：

```
[silu-inplace-2d-64x64] ascendc max_diff=2.730469
[silu-inplace-2d-64x64] pto max_diff=3.519531
（100% 元素不匹配）

非原地（独立 dst/src buffer）双后端均正确。
```

## Expected Behavior

`T.tile.silu(a_ub, a_ub)` 原地运算应输出与 `torch.nn.functional.silu(a)` 一致的结果（与 `T.tile.relu` / `T.tile.leaky_relu` 原地行为一致）；若不支持原地，应在编译期或运行时报错，而非静默产生错误结果。

## Actual Behavior

原地运算静默产生错误结果：64×64 float16 实测 ascendc max_diff≈2.7、pto ≈3.5，100% 元素不匹配，无任何报错。

**根因已定位**：CANN `silu_common_impl.h::SiluCalcSimplified` 的指令序列：

```cpp
Muls(dst, src, -1);         // step1: dst = -src        ← in-place 时 src 被覆盖
Exp(dst, dst);              // step2: dst = e^(-src)
Adds(dst, dst, 1.0);        // step3: dst = 1+e^(-src)
Div(dst, srcAddr, dstAddr); // step4: dst = src/(1+e^(-src))  ← 依赖原始 src，已被 step1 破坏
```

step4 的 `Div` 需要原始 src，但 step1 已将其覆盖为 `-src` → 数值错误。

**对照**：`sigmoid_impl.h::SigmoidIntrinsicsImpl` 最后一步为 `Div(dst, stackBuffer, dst)`，分母来自独立临时 buffer、不依赖原始 src，故 `T.tile.sigmoid` 原地在 ascendc 正确（pto 仍错误，见附注）。

## Additional Context

**Programming Mode:**
- [x] Developer mode (automatic sync/buffer management)
- [ ] Expert mode (manual T.Scope/T.barrier_all)
- [ ] Mixed mode

**Backend:**
- [x] Ascend C & PTO backend
- [ ] AscendNPU IR backend

**Pass Configs (if applicable):**
```python
pass_configs = {
    tilelang.PassConfigKey.TL_ASCEND_AUTO_CV_COMBINE: True,
    tilelang.PassConfigKey.TL_ASCEND_AUTO_CV_SYNC: True,
    tilelang.PassConfigKey.TL_ASCEND_AUTO_SYNC: True,
    tilelang.PassConfigKey.TL_ASCEND_MEMORY_PLANNING: True,
}
```

**Operator Type:**
激活函数（SiLU / element-wise unary）

**附注（关联问题）**：
- `T.tile.sigmoid` 原地：ascendc 正确（max_diff≈5e-4），pto 错误（max_diff≈0.98）——同样建议模板层对 dst==src 别名做拷贝保护或编译期拒绝。
- 建议修复方向：CANN 侧修复 `SiluCalcSimplified`（先拷贝 src 或重排为使用独立临时 buffer 的序列）；TileLang 模板层对齐名（dst==src）做拷贝保护或编译期拒绝。

**Additional Information:**
- Have you checked the [Programming Guide](../docs/TileLang-Ascend%20Programming%20Guide.md)? Yes
- Have you searched existing issues? Yes
- Any other context or screenshots? 详细记录见 PR #1657 对应更新说明与 `request/说明和审查/框架.md` 第 12 条

### PR DESCRIPTION
**Pull request by fredrekelthen**
## 问题背景

Fixes #1658

`T.tile.silu` 原地运算（`dst` 与 `src` 为同一 buffer）时**静默产生错误结果**，ascendc / pto 双后端均受影响，无任何编译期/运行期报错。issue 附注的 `T.tile.sigmoid` 原地问题则为 ascendc 正确、pto 错误。

真机复现（910B3，64×64 float16 随机输入）：

| 场景 | target | 修复前 max_diff | 修复后 max_diff |
|---|---|---|---|
| silu 原地（dst==src） | ascendc | 2.21（100% 元素错误） | 0.001953 ✓ |
| silu 原地（dst==src） | pto | 3.67（100% 元素错误） | 0.001953 ✓ |
| sigmoid 原地（附注 bug） | pto | 0.977 | 0.000488 ✓ |
| sigmoid 原地 | ascendc | 0.000488（本就正确） | 0.000488（不回归） |

## 根因分析

**ascendc 后端**：codegen 直接透传 `AscendC::Silu`，CANN 2201 路径（`SiluCalcSimplified`）的指令序列原地不安全：

```cpp
Muls(dst, src, -1);          // step1: dst = -src        ← 原地时 src 已被覆盖
Exp(dst, dst);
Adds(dst, dst, 1.0);
Div(dst, srcAddr, dstAddr);  // step4: 依赖原始 src → 已被 step1 破坏
```

对照 devkit 的 `SigmoidIntrinsicsImpl`：末步 `Div(dst, stackBuffer, dst)` 的分子来自独立临时 buffer，不依赖原始 src，故 sigmoid 原地在 ascendc 正确——**"分子来自独立 buffer、dst 别名分母"是本次修复采用的安全模式**。

**pto 后端**：两层问题——
1. `TSIGMOID` 把中间量写进 `src0`（篡改用户只读输入，即 #1517 列出的契约破坏）；
2. `TSILU` 经 `TRECIP(dst==src)` 实现，与既有 `TSIGMOID` 组合后末步 `TMUL(dst, tmp, dst)` 读到的是已篡改的中间量。

## 修复方案（三层）

| 文件 | 改动 |
|---|---|
| `src/tl_templates/pto/common.h` | **TSILU** 重写为 `TMOV(tmp,src) → TMULS(dst,src,-1) → TEXP → TADDS → TDIV(dst,tmp,dst)`（消除 ISA 级不安全的 `TRECIP(dst==src)`，中间量只写 dst/tmp，src 只读）；**TSIGMOID** 增加 tmp 参数，中间量全走 tmp（修复原地错误 + 不再篡改用户 src） |
| `src/target/codegen_ascend_pto.cc` | `SigmoidCodegen` 仿 `SiluCodegen` 分配内部 tmp 并推进 `max_ub_addr_`（32B 对齐，防地址重叠） |
| `tilelang/language/ascend_tile.py` | `silu()` 增加别名检测（`_is_aliased`，覆盖 BufferRegion 同 buffer 重叠）→ 自动插入 scratch buffer 并以 mul-by-one 拷贝保护（不用 `T.copy` 是因 auto-sync pass 在某些序列下会丢失 `copy_ub_to_ub` 的屏障，属另一既有问题）；docstring 更新原地语义 |

ascendc 后端不依赖 CANN 修复（上游 devkit 2201 路径仍是不安全序列），由 Python 层别名保护兜底：原地时先 `Muls(tmp, src, 1.0)` 复制到 scratch，再对无别名的 (dst, tmp) 调用 `AscendC::Silu`。

## 验证结果（真机 910B3，CANN 8.5.2）

- 新增测试 `test_silu_inplace` / `test_sigmoid_inplace`（float/float16 × ascendc/pto × 256²/512²）：**16/16 PASS**
- 存量回归：`-k "silu or sigmoid or relu or exp or activation or sin or cos"` 85/85 PASS；`test_tilelang_ascend_language_explicit_tmp.py` 68/68 PASS（含 TSIGMOID 源码断言）
- 附加专项验证（未入库）：region 同 buffer 重叠、链式两次原地、src 输入保持性（#1517 契约）、±inf/NaN/±0/dtype 边界 IEEE 语义、threads=2 拆分、1D/边缘 shape——全部通过
- ruff check/format + clang-format 全绿

## 已知遗留（与本次修复无关的既有问题，建议后续立项）

1. pto codegen 的 `SiluCodegen`/`SigmoidCodegen` tmp 分配不检查 196608B UB 上限，同一 kernel 多次大 tile 调用会溢出（两次非别名 128×128 f32 silu 即可复现）；
2. auto-sync pass 在 `copy_ub_to_ub` 后跟某些 V 操作时丢失整条链路的 SetFlag/WaitFlag。

基线：基于最新 `ascendc_pto`（cd38d3f3），无冲突。

### CODE REVIEW

File: src/target/codegen_ascend_pto.cc [hunk 1]
```diff
@@ -3436,11 +3436,27 @@ void CodeGenTileLangAscendPto::SigmoidCodegen(const CallNode *op,
...
   std::string src_name = ResolveUbSliceName(src_shape_info);
   std::string dst_name = ResolveUbSliceName(dst_shape_info);
+  std::string tmp_name =
+      GetTempVarName(dst_shape_info.ub_name) + "_sigmoid_tmp";
 
+  // Update max_ub_addr_ after allocating temporary buffer
+  int32_t elem_bytes = GetTypeLen(dst_shape_info.type);
+  int64_t tmp_buffer_size = tpl_row * tpl_col * elem_bytes;
+  int64_t tmp_addr = max_ub_addr_; // Save original address before alignment
+  max_ub_addr_ += tmp_buffer_size;
```
Comment (lwwangcgz) [inline src/target/codegen_ascend_pto.cc]: [P1] Keep this scratch inside memory planning

This grows `max_ub_addr_` during codegen, after `AscendMemoryPlanning`, while `GetPTOWorkspaceSpec` still classifies `tl.ascend_sigmoid` as `NoWorkspace`. The planner therefore cannot capacity-check or reuse it: a normal out-of-place 128x128 float sigmoid uses src + dst + tmp = 196,608 B, exceeding the A2/A3 196,352-B UB limit. Please inject/consume a planned workspace (or compute through `dst`) instead of appending hidden storage here.

File: tilelang/language/ascend_tile.py [hunk 2]
```diff
@@ -1187,7 +1230,13 @@ def silu(dst: Buffer | BufferRegion, src: Buffer | BufferRegion):
     Note:
```
Comment (lwwangcgz) [inline tilelang/language/ascend_tile.py]: [P2] Check whether the regions actually overlap

`same_as` only proves that two operands share a backing `Buffer`; it does not mean their `BufferRegion` ranges overlap. For example, `silu(buf[0:64, :], buf[64:128, :])` is disjoint, but this returns true and allocates a full clone. On a 128x320 float UB that turns 163,840 B into 245,760 B. Please compare byte intervals when possible and copy only when overlap exists or cannot be ruled out.

File: src/target/codegen_ascend_pto.cc [hunk 3]
```diff
@@ -3436,11 +3436,27 @@ void CodeGenTileLangAscendPto::SigmoidCodegen(const CallNode *op,
 
   std::string src_name = ResolveUbSliceName(src_shape_info);
   std::string dst_name = ResolveUbSliceName(dst_shape_info);
+  std::string tmp_name =
+      GetTempVarName(dst_shape_info.ub_name) + "_sigmoid_tmp";
 
+  // Update max_ub_addr_ after allocating temporary buffer
```
Comment (ChaoyangJi) [inline src/target/codegen_ascend_pto.cc]: Sigmoid scratch会累加并越过UB

PTO 每次调用都从 max_ub_addr_ 新分配完整 tile，既不复用也不检查 196352 B 的 planner 上限。两个 (128,128) float32 buffer 连续调用两次 sigmoid 时，第二块 scratch 位于 [196608,262144)，确定越界。

File: tilelang/language/ascend_tile.py [hunk 4]
```diff
@@ -1187,7 +1230,13 @@ def silu(dst: Buffer | BufferRegion, src: Buffer | BufferRegion):
...
         - Supports data types: half, float (Atlas A2/A3)
         - SiLU = x * sigmoid(x) = x / (1 + exp(-x))
+        - In-place use (``dst`` is the same buffer as ``src``) is supported:
+          the aliased source is transparently copied to a scratch UB buffer
+          before the computation, since the final division needs the original
+          source values after dst has been overwritten.
     """
+    if _is_aliased(dst, src):
```
Comment (ChaoyangJi) [inline tilelang/language/ascend_tile.py:1263]: 原地 SiLU 同时占用三份 tile
前端创建 alias 副本，PTO codegen 又创建内部 scratch。PR 测试中的 (128,128) float32 已占 65536 × 3 = 196608 B；再有任何存活 UB buffer 就会越过物理容量。

File: tilelang/language/ascend_tile.py [hunk 5]
```diff
@@ -1147,6 +1148,44 @@ def exp(dst: Buffer | BufferRegion, src0: Buffer | BufferRegion):
     return unary_op(dst, src0, "exp")
 
 
+def _underlying_buffer(br: Buffer | BufferRegion) -> Buffer:
+    return br.buffer if isinstance(br, BufferRegion) else br
+
+
+def _is_aliased(dst: Buffer | BufferRegion, src: Buffer | BufferRegion) -> bool:
```
Comment (ChaoyangJi) [inline tilelang/language/ascend_tile.py:1161]: 不相交 BufferRegion 被误判为 alias

当前仅比较底层 Buffer 身份。同一 Buffer 中静态不相交的两个 slice 也会触发完整复制，可能让原本合法的近容量 kernel 编译失败，并增加一次向量遍历。

File: src/target/codegen_ascend_pto.cc (hunk 3 above)
Comment (fredrekelthen) [reply src/target/codegen_ascend_pto.cc]: 已在 04803c33 修复：

- 新增 `AllocActivationScratch` 共享槽：silu/sigmoid 复用同一 scratch 地址（实测 3 次 (128,128) float32 sigmoid 的 scratch 地址均为 131072，不再逐次累加）；
- 新增 196608B（192KiB，实测 UB 包络）容量检查：scratch 会越界时编译期直接报 `exceeds the UB capacity` 附带可用信息（planned buffers 结束地址、建议减小 tile / 减少激活算子），实测 (128,160)×2 场景正确拒绝，临界预算场景（2×65536 planned + 65536 scratch = 196608）正常编译运行。

reviewer 场景本身（两次连续 sigmoid）现在通过：scratch 只占一份 65536B。

File: tilelang/language/ascend_tile.py (hunk 4 above)
Comment (fredrekelthen) [reply tilelang/language/ascend_tile.py:1263]: 已在 04803c33 修复：新增 `TSiluNoAlias` 免 scratch 快速路径。

codegen 按 buffer 名判定：dst/src 为**不同 buffer** 时（memory planner 保证同时活跃的 buffer 地址不重叠，前端 alias 副本与 dst 在调用点同时活跃），生成 `TMULS(dst,src,-1) → TEXP(dst,dst) → TADDS(dst,dst,1) → TDIV(dst,src,dst)`——src 只读、dst 只别名分母（与 devkit `SigmoidIntrinsicsImpl` 的 Div 安全模式同构），**无内部 scratch**。dst/src 同名（同 buffer 任意切片）时仍走带 TMOV 的安全序列。

效果：pto 原地 silu 总占用从 3 份降到 **2 份**（与非原地等价）。实测 (128,128) float32 原地生成代码仅 2 个 TASSIGN（a_ub + 前端副本）、无 `_silu_tmp`；原地/链式大 tile 场景（修复前 aicore 崩溃）现在双后端通过。

前端副本保留的原因：ascendc 的 CANN `Silu`（2201 路径 `SiluCalcSimplified`）原地不安全，需要该副本兜底；最新 devkit 该路径仍未修复。

File: tilelang/language/ascend_tile.py (hunk 5 above)
Comment (fredrekelthen) [reply tilelang/language/ascend_tile.py:1161]: 已在 04803c33 修复：`_is_aliased` 增加静态判别。

规则：不同 buffer → 恒非别名（planner 保证）；同 buffer 时，dst/src 均为 BufferRegion 且**任一维度**的常量区间可证不相交（`d.min + d.extent <= s.min || s.min + s.extent <= d.min`，均为 IntImm）→ 非别名，不做拷贝；区间非常量或不可证 → 保守按别名处理（正确性优先）。

实测：`silu(a[0:32,:], a[32:64,:])`（同 buffer 不相交）无拷贝直通；`silu(a[0:32,:], a[16:48,:])`（部分重叠）仍受保护——两种情况双后端数值均正确。codegen 侧同步收紧：仅 dst/src 同名 buffer 才走带 scratch 的安全序列，异名（含同 buffer 不相交 region lowering 后的情形）走快速路径。

**Review by ChaoyangJi — APPROVED** [review]
lgtm
